### PR TITLE
Streamline Regex path to matching, and improve Replace/Split

### PR DESCRIPTION
**Pull request by stephentoub**
I spent some time looking at costs in RegexRedux.  This PR makes two primary changes:

1. Streamlines the path from the public APIs to the point where we get a match.  This is primarily about reducing unnecessary checks, coding to the common case, etc.  The impact can be best seen by looking at some of your checked-in benchmarks:

|          Method |           Toolchain |              Options |      Mean | Ratio |
|---------------- |-------------------- |--------------------- |----------:|------:|
|    Date_IsMatch | \master\corerun.exe |                 None | 191.06 ns |  1.00 |
|    Date_IsMatch |     \pr\corerun.exe |                 None | 180.77 ns |  0.95 |
|                 |                     |                      |           |       |
| Date_IsNotMatch | \master\corerun.exe |                 None | 433.70 ns |  1.00 |
| Date_IsNotMatch |     \pr\corerun.exe |                 None | 421.94 ns |  0.97 |
|                 |                     |                      |           |       |
|    Date_IsMatch | \master\corerun.exe |             Compiled |  72.29 ns |  1.00 |
|    Date_IsMatch |     \pr\corerun.exe |             Compiled |  62.84 ns |  0.87 |
|                 |                     |                      |           |       |
| Date_IsNotMatch | \master\corerun.exe |             Compiled | 121.41 ns |  1.00 |
| Date_IsNotMatch |     \pr\corerun.exe |             Compiled | 118.23 ns |  0.97 |
|                 |                     |                      |           |       |
|    Date_IsMatch | \master\corerun.exe | IgnoreCase, Compiled |  92.62 ns |  1.00 |
|    Date_IsMatch |     \pr\corerun.exe | IgnoreCase, Compiled |  80.38 ns |  0.87 |
|                 |                     |                      |           |       |
| Date_IsNotMatch | \master\corerun.exe | IgnoreCase, Compiled | 174.41 ns |  1.00 |
| Date_IsNotMatch |     \pr\corerun.exe | IgnoreCase, Compiled | 168.18 ns |  0.96 |

2. Fixing our Regex.Replace implementation (and also Regex.Split, as it shared some common internal functionality).  There are two issues here.  First, Replace works by repeatedly calling Match, but that means every time we find a match as part of the replace, we're a) creating a new Match object, and b) having to re-enter the main engine.  We can instead internally use a callback model where we a) reuse the same Match object over and over, and b) never leave the main engine loop during the whole replace operation.  This helps, but it's not the most impactful piece.  Second, Replace needs to build up the result string to be returned.  Previously it was using a StringBuilder, but as that ends up allocating a StringBuilder object and the underlying char[], some time ago it was changed to use ValueStringBuilder.  The problem with this is that whereas StringBuilder is implemented as a linked-list of StringBuilder instances, explicitly avoiding creating super long arrays, ValueStringBuilder is implemented around a single large array, which means every time it grows, it needs to get a new array, copy the old to the new, etc.  It gets these arrays from the ArrayPool, which generally works ok as long as it actually can get such an array, but when the input fed to replace is very large, at some point ArrayPool might not store arrays large enough to satisfy the request, and it'll just allocate a new array.  RegexRedux is feeding in 50MB text file, which means we end up allocating huge arrays as part of each Regex.Replace.  I changed the implementation to not use ValueStringBuilder, and instead just use a simple struct builder that maintains an array of segments rather than individual characters; while in an extreme corner case we could end up replacing each individual character in the input and end up with a larger array than we previously had, by far the most common use case is at the other extreme, with few or larger replacements, in which case we end up with much less memory usage.  We then also avoid any copying of text while building up the segments, and instead just copy directly into the resulting string as part of the final string.Create.

```
d:\coreclrtest\master>corerun C:\Users\stoub\source\repos\RegexRedux\bin\Release\netcoreapp3.1\RegexRedux.dll
00:00:03.1093741        4,167,418,952.00
00:00:02.8592129        4,162,813,888.00
00:00:02.7930212        4,162,870,888.00
00:00:02.8526279        4,162,852,040.00
00:00:02.7953997        4,162,841,840.00
00:00:02.8235776        4,162,830,816.00
00:00:02.7939749        4,162,816,616.00
00:00:02.8719042        4,162,811,056.00
00:00:02.7970820        4,162,838,744.00
00:00:02.8178119        4,162,815,960.00

d:\coreclrtest\pr>corerun C:\Users\stoub\source\repos\RegexRedux\bin\Release\netcoreapp3.1\RegexRedux.dll
00:00:02.4166629        1,196,928,608.00
00:00:02.1559750        1,162,969,456.00
00:00:02.1224511        1,163,039,064.00
00:00:02.0927512        1,162,969,600.00
00:00:01.9971182        1,129,376,872.00
00:00:02.1589007        1,129,425,264.00
00:00:02.0703367        1,129,411,112.00
00:00:02.1120507        1,162,966,704.00
00:00:02.0825877        1,163,023,024.00
00:00:02.1078391        1,129,439,728.00

d:\coreclrtest\pr>
```

I also cleaned up some style along the way.

cc: @danmosemsft, @eerhardt, @pgovind, @ViktorHofer

### CODE REVIEW

File: src/libraries/System.Text.RegularExpressions/src/System/Text/SegmentStringBuilder.cs [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+using System.Buffers;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace System.Text
+{
+    /// <summary>Provides a value type string builder composed of individual segments represented as <see cref="ReadOnlyMemory{T}"/> instances.</summary>
+    [DebuggerDisplay("Count = {_count}")]
+    internal struct SegmentStringBuilder
```
Comment (danmosemsft) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/SegmentStringBuilder.cs]: Probably worth adding a few tests?

Comment (stephentoub) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/SegmentStringBuilder.cs]: Doesn't seem valuable enough to me; we don't have tests for most other internal implementation details.  It's just a helper for the Replace implementation; if it's buggy in a way that impacts Replace, that'll show up in tests for Replace.

File: src/libraries/System.Text.RegularExpressions/src/System/Text/SegmentStringBuilder.cs [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+                {
+                    segment.Span.CopyTo(dest);
+                    dest = dest.Slice(segment.Length);
+                }
+            });
+
+            ReadOnlyMemory<char>[] array = _array;
+            AsSpan().Clear(); // clear just what's been filled
+            this = default;
+            ArrayPool<ReadOnlyMemory<char>>.Shared.Return(array);
```
Comment (danmosemsft) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/SegmentStringBuilder.cs]: `_array = null;` ?

Comment (stephentoub) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/SegmentStringBuilder.cs]: The `this = default;` is doing that.

Comment (danmosemsft) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/SegmentStringBuilder.cs]: Oh, I didn't acutally know you could do that.

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexRunner.cs [hunk 3]
```diff
@@ -85,118 +85,251 @@ public abstract class RegexRunner
...
-            bump = runregex.RightToLeft ? -1 : 1;
-            stoppos = runregex.RightToLeft ? runtextbeg : runtextend;
-
-            runtextpos = textstart;
+            // Handle timeout argument
+            _timeout = -1; // (int)Regex.InfiniteMatchTimeout.TotalMilliseconds
+            bool ignoreTimeout = _ignoreTimeout = Regex.InfiniteMatchTimeout == timeout;
+            if (!ignoreTimeout)
+            {
+                // We are using Environment.TickCount and not Timewatch for performance reasons.
```
Comment (danmosemsft) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexRunner.cs]: tpyo: "Timewatch" ?

Comment (stephentoub) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexRunner.cs]: Thanks, fixed.

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/ThrowHelper.cs [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+                ExceptionResource.LengthNotNegative => SR.LengthNotNegative,
+                ExceptionResource.OnlyAllowedOnce => SR.OnlyAllowedOnce,
+                ExceptionResource.ReplacementError => SR.ReplacementError,
+                _ => null
+            };
+    }
+
+    internal enum ExceptionArgument
+    {
+        assemblyname,
```
Comment (eerhardt) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/ThrowHelper.cs]: (nit) any reason why these enum values are camel cased?

Comment (stephentoub) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/ThrowHelper.cs]: To exactly match the arguments they represent.  We do the same thing in other ThrowHelpers, e.g.
https://github.com/dotnet/runtime/blob/c26a65343448771e0dcdbcfdd5756c516a919e6b/src/libraries/System.Private.CoreLib/src/System/ThrowHelper.cs#L853-L942
https://github.com/dotnet/runtime/blob/c26a65343448771e0dcdbcfdd5756c516a919e6b/src/libraries/System.Memory/src/System/ThrowHelper.cs#L108-L124
https://github.com/dotnet/runtime/blob/c26a65343448771e0dcdbcfdd5756c516a919e6b/src/libraries/System.Linq/src/System/Linq/ThrowHelper.cs#L61-L80

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/ThrowHelper.cs [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+                ExceptionArgument.length => "length",
+                ExceptionArgument.matchTimeout => "matchTimeout",
+                ExceptionArgument.name => "name",
+                ExceptionArgument.options => "options",
+                ExceptionArgument.pattern => "pattern",
+                ExceptionArgument.regexinfos => "regexinfos",
+                ExceptionArgument.replacement => "replacement",
+                ExceptionArgument.startat => "startat",
+                ExceptionArgument.str => "str",
+                ExceptionArgument.value => "value",
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/ThrowHelper.cs]: These should all be nameof.  I'll fix that.

File: src/libraries/System.Text.RegularExpressions/src/System/Text/SegmentStringBuilder.cs [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+            Array.Copy(array, newArray, _count);
+            ArrayPool<ReadOnlyMemory<char>>.Shared.Return(array, clearArray: true);
+            newArray[_count++] = segment;
+        }
+
+        /// <summary>Gets a span of all segments in the builder.</summary>
+        /// <returns></returns>
+        public Span<ReadOnlyMemory<char>> AsSpan() => new Span<ReadOnlyMemory<char>>(_array, 0, _count);
+
+        /// <summary>Creates a string from all the segments in the builder and then disposes of the builder.</summary>
```
Comment (eerhardt) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/SegmentStringBuilder.cs]: > and then disposes of the builder

Is this a typical pattern for `ToString()` to mutate/dispose the instance? I think I would find that behavior surprising if I used this type.

Comment (stephentoub) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/SegmentStringBuilder.cs]: On public types, probably not.  But it's what we do in ValueStringBuilder:
https://github.com/dotnet/runtime/blob/c26a65343448771e0dcdbcfdd5756c516a919e6b/src/libraries/Common/src/System/Text/ValueStringBuilder.cs#L86-L91

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.Split.cs [hunk 7]
```diff
@@ -66,94 +72,79 @@ public string[] Split(string input, int count, int startat)
...
             {
-                result = new string[1];
-                result[0] = input;
-                return result;
+                return new[] { input };
             }
 
-            count -= 1;
+            count--;
+            var state = (al: new List<string>(), prevat: 0, input, count);
```
Comment (eerhardt) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.Split.cs]: I know that the original variable was named the same, but what does `al` stand for? The only thing I can guess is "a list".

Comment (stephentoub) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.Split.cs]: Heh, good question, no idea.  I'll change it.

**Review by eerhardt — APPROVED** [review]
LGTM. Nice work, @stephentoub.

**Comment by stephentoub** [comment]
Thanks, Eric.

File: src/libraries/System.Text.RegularExpressions/src/System/Text/SegmentStringBuilder.cs (hunk 1 above)
Comment (pgovind) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/SegmentStringBuilder.cs]: I think `ReadOnlySequence` from System.Memory does something very similar to `SegmentStringBuilder`?

Comment (stephentoub) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/SegmentStringBuilder.cs]: Not without allocating for each segment.

**Review by pgovind — APPROVED** [review]
Nice!

**Pull request merged**